internal/exec/stages/files/units_test.go

```diff
@@ -0,0 +1,68 @@
+// Copyright 2020 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package files
+
+import (
+	"testing"
+
+	"github.com/coreos/ignition/v2/config/shared/errors"
+	"github.com/coreos/ignition/v2/config/v3_1_experimental/types"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestParseInstanceUnit(t *testing.T) {
+	type in struct {
+		unit types.Unit
+	}
+	type out struct {
+		serviceInstance string
+		instance        string
+		parseErr        error
+	}
+	tests := []struct {
+		in  in
+		out out
+	}{
+		{in: in{types.Unit{Name: "echo@bar.service"}},
+			out: out{serviceInstance: "echo@.service", instance: "bar",
+				parseErr: nil},
+		},
+
+		{in: in{types.Unit{Name: "echo@foo.service"}},
+			out: out{serviceInstance: "echo@.service", instance: "foo",
+				parseErr: nil},
+		},
+		{in: in{types.Unit{Name: "echo@fooservice"}},
+			out: out{serviceInstance: "", instance: "",
+				parseErr: errors.ErrNoDotSuffix},
+		},
+		{in: in{types.Unit{Name: "echo@.service"}},
+			out: out{serviceInstance: "echo@.service", instance: "",
+				parseErr: nil},
+		},
+		{in: in{types.Unit{Name: "postgresql@9.3-main.service"}},
+			out: out{serviceInstance: "postgresql@.service", instance: "9.3-main",
+				parseErr: nil},
+		},
+	}
+	for i, test := range tests {
+		serviceInstance, instance, err := parseInstanceUnit(test.in.unit)
+		if (test.out.parseErr != nil) && test.out.parseErr != err {
```

Comment:
If the function returns an unexpected error, we won't report the error, only the surprising output.

Reply:
I'm thinking about a case where I will encounter an unexpected error.
```
func parseInstanceUnit(unit types.Unit) (string, string, error) {
	at := strings.Index(unit.Name, "@")
	dot := strings.LastIndex(unit.Name, ".")
	if dot == -1 {
		return "", "", errors.ErrNoSystemdExt
	}
	instance := unit.Name[at+1 : dot]
	serviceInstance := unit.Name[0:at+1] + unit.Name[dot:len(unit.Name)]
	return serviceInstance, instance, nil
}
```
In this function, we're expecting at least one `@` in a `unit.name` string(https://github.com/coreos/ignition/pull/934/files#r395010754) so not handling any error there. The code is already catching the error if the index of `dot` is `-1`.
Could you please provide more explanation here?

Reply:
I was referring to the test harness.  If we run a test that expects success, and instead receive an error, we don't actually report the contents of the error.

Reply:
Fixed.

Reply:
This is not actually fixed.  What is the purpose of the `test.out.parseErr != nil` check?

Reply:
I see what you're saying. I have removed that part from the code.